### PR TITLE
alphametics: add exercise with generator

### LINKED ISSUE #917: alphametics: implement exercise
[Specification](https://github.com/exercism/problem-specifications/tree/master/exercises/alphametics)

### PR DESCRIPTION
**Pull request by leenipper**
Resolves #917

### CODE REVIEW

**Comment by leenipper** [comment]
I used a simple brute-force algorithm. Not the fastest, but good enough for the example.

File: exercises/alphametics/.meta/gen.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+func (c OneCase) SortedMapString() string {
+	letters := make([]string, 0, len(c.Expected))
+	for s, _ := range c.Expected {
+		letters = append(letters, s)
+	}
+	result := ""
+	sort.Strings(letters)
+	for _, s := range letters {
+		result += fmt.Sprintf(`"%s": %d, `, s, c.Expected[s])
```
Comment (ferhatelmas) [inline exercises/alphametics/.meta/gen.go]: Might generate strings directly and sort them and finally join by `, `

File: exercises/alphametics/example.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	nLetters     int
+}
+
+func Solve(puzzle string) (map[string]int, error) {
+	p := parsePuzzle(puzzle)
+	if p == nil {
+		return nil, errors.New("invalid puzzle")
+	}
+	r, err := solvePuzzle(p)
+	return r, err
```
Comment (ferhatelmas) [inline exercises/alphametics/example.go]: just `return solvePuzzle(p)`

File: exercises/alphametics/example.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+	for v := 0; v < len(p.letterValues); v++ {
+		if p.letterValues[v] == -1 {
+			p.lettersUsed[n] = rune(v) // 0 == 'A'
+			n++
+		}
+	}
+	return p
+}
+
+func solvePuzzle(p *problem) (map[string]int, error) {
```
Comment (ferhatelmas) [inline exercises/alphametics/example.go]: might be a method of problem

File: exercises/alphametics/example.go [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+		}
+	}
+	return true
+}
+
+// puzzleMap creates a "by letter" map from the letterValues used.
+func puzzleMap(p *problem) map[string]int {
+	pm := make(map[string]int, p.nLetters)
+	for _, v := range p.lettersUsed {
+		r := rune(v) + 'A'
```
Comment (ferhatelmas) [inline exercises/alphametics/example.go]: rune conversion is unnecessary

File: exercises/alphametics/example.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+		for d, r := range valueStrings[i] {
+			j := len(valueStrings[i]) - 1 - d
+			// Each vDigits value is 0 for nothing in this position,
+			// or 1..26 for 'A'..'Z'.
+			p.vDigits[i][j] = r - 'A' + 1
+		}
+	}
+	// Make a list of the letters used, where 0 == 'A'
+	p.lettersUsed = make([]rune, p.nLetters)
+	n := 0
```
Comment (ferhatelmas) [inline exercises/alphametics/example.go]: since it's only used in for, might define in the beginning of for.

**Comment by leenipper** [comment]
Thanks @ferhatelmas !

**Pull request merged**